altos-rust/port/cortex-m0/src/peripheral/usart/mod.rs

```diff
@@ -1,123 +1,305 @@
 pub use self::baudr::BaudRate;
 
-#[derive(Copy, Clone)]
+pub const USART2_TX_BUFFER_FULL_CHAN: usize = 43;
+pub const USART2_TC_CHAN: usize = 43 * 2;
+
+/// STM32F0 has two Usart registers available.
+#[derive(Copy, Clone, Debug)]
 pub enum UsartX {
     Usart1,
     Usart2,
 }
 
-#[derive(Copy, Clone)]
+/// Defines the memory address for the Usart and each variable needed
+/// to access the registers that comprise the Usart.
+/* mem_addr: Memory address of the Usart
+ * control: Collection of control registers
+ * baud: Baud rate register
+ * tdr: Transmit data register
+ * rdr: Read data register
+ * isr: Interrupt service register
+ * icr: Interrupt clear register
+ */
+#[derive(Copy, Clone, Debug)]
 pub struct Usart {
     mem_addr: *const u32,
     control: UsartControl,
     baud: BRR,
     tdr: TDR,
+    rdr: RDR,
     isr: ISR,
+    icr: ICR,
 }
 
+/// Wraps the Usart memory address in order to make it volatile.
 impl Control for Usart {
     unsafe fn mem_addr(&self) -> Volatile<u32> {
         Volatile::new(self.mem_addr as *const u32)
     }
 }
 
+/// Function implementations for the Usart struct. These are the wrapper
+/// functions that can be called on a Usart object in order to set
+/// specific bits in each of the registers contained within the Usart struct.
+/// Each function calls the appropriate function associated with the
+/// register needed to set the correct bits.
 impl Usart {
+
+    /// Creates a new Usart. Pattern matches the argument passed in, and
+    /// creates and initializes either a new Usart1 or a new Usart2.
+    /// The base address of the Usart being used is passed into each 'new` call
+    /// the proper address with offset can be calculate for the associated
+    /// register.
     pub fn new(x: UsartX) -> Self {
         match x {
             UsartX::Usart1 => Usart {
                 mem_addr: USART1_ADDR,
                 control: UsartControl::new(USART1_ADDR),
                 baud: BRR::new(USART1_ADDR),
                 tdr: TDR::new(USART1_ADDR),
+                rdr: RDR::new(USART1_ADDR),
                 isr: ISR::new(USART1_ADDR),
+                icr: ICR::new(USART1_ADDR),
             },
             UsartX::Usart2 => Usart {
                 mem_addr: USART2_ADDR,
                 control: UsartControl::new(USART2_ADDR),
                 baud: BRR::new(USART2_ADDR),
                 tdr: TDR::new(USART2_ADDR),
+                rdr: RDR::new(USART2_ADDR),
                 isr: ISR::new(USART2_ADDR),
+                icr: ICR::new(USART2_ADDR),
             },
         }
     }
 
-    // TODO: Change &self -> &mut self for all these?? Maybe??
-    pub fn enable_usart(&self) {
+    // --------------------------------------------------------------
+    /// This section pertains to calls made to the control registers.
+    /// See control.rs for more specific information.
+
+    /// Makes call using the Usart control variable to enable the Usart.
```

Comment:
This is where we want to explain the meaning of each function call, we can probably propagate up each doc comment from the submodules to here. This applies to all the wrapper functions.